File: DSA/Ayush_Mishra_590014844/Day_51_Question_3.py
def min_window_substring(s, t):
    if not s or not t:
        return ""

    freq_t = {}
    for ch in t:
        freq_t[ch] = freq_t.get(ch, 0) + 1

    freq_s = {}
    have, need = 0, len(freq_t)
    left = 0
    res = [-1, -1]
    res_len = float('inf')

    for right in range(len(s)):
        ch = s[right]
        freq_s[ch] = freq_s.get(ch, 0) + 1

        if ch in freq_t and freq_s[ch] == freq_t[ch]:
            have += 1

        while have == need:
            if right - left + 1 < res_len:
                res = [left, right]
                res_len = right - left + 1

            freq_s[s[left]] -= 1
            if s[left] in freq_t and freq_s[s[left]] < freq_t[s[left]]:
                have -= 1
            left += 1

    return s[res[0]:res[1]+1] if res_len != float('inf') else ""

s = input().strip()
t = input().strip()
print(min_window_substring(s, t))

File: DSA/Ayush_Mishra_590014844/Day_51_Question_3.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

int minWindow(char *s, char *t, char *result) {
    int ns = strlen(s), nt = strlen(t), freqT[256] = {0}, freqS[256] = {0};
    for (int i = 0; i < nt; i++)
        freqT[t[i]]++;

    int left = 0, count = 0, minLen = INT_MAX, start = -1;
    for (int right = 0; right < ns; right++) {
        freqS[s[right]]++;
        if (freqT[s[right]] && freqS[s[right]] <= freqT[s[right]])
            count++;
        while (count == nt) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                start = left;
            }
            freqS[s[left]]--;
            if (freqT[s[left]] && freqS[s[left]] < freqT[s[left]])
                count--;
            left++;
        }
    }
    if (start == -1)
        return 0;
    strncpy(result, s + start, minLen);
    result[minLen] = '\0';
    return minLen;
}

int main() {
    char s[1000], t[1000], res[1000];
    scanf("%s %s", s, t);
    if (minWindow(s, t, res))
        printf("%s\n", res);
    else
        printf("\n");
    return 0;
}
